python/tritonserver/_c/tritonserver_pybind.cc

```diff
@@ -896,107 +871,120 @@ class PyInferenceRequest
+    //       static initialized?
+    static std::once_flag allocator_init_once;
+    std::call_once(allocator_init_once, [this]() {
+      struct TRITONSERVER_ResponseAllocator* allocator_raw = nullptr;
+      ThrowIfError(TRITONSERVER_ResponseAllocatorNew(
+          &allocator_raw, ResponseAllocatorAllocFn, ResponseAllocatorReleaseFn,
+          nullptr /* start_fn */));
+      allocator_.reset(allocator_raw);
+      ThrowIfError(TRITONSERVER_ResponseAllocatorSetQueryFunction(
+          allocator_.get(), ResponseAllocatorQueryFn));
+      ThrowIfError(TRITONSERVER_ResponseAllocatorSetBufferAttributesFunction(
+          allocator_.get(), ResponseAllocatorBufferAttributesFn));
+    });
+  }
+  static TRITONSERVER_Error* ResponseAllocatorQueryFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* userp,
+      const char* tensor_name, size_t* byte_size,
+      TRITONSERVER_MemoryType* memory_type, int64_t* memory_type_id)
   {
-    request_callback_resource_.reset(
-        new CallbackResource(release, user_object));
-    ThrowIfError(TRITONSERVER_InferenceRequestSetReleaseCallback(
-        triton_object_, PyTritonRequestReleaseCallback,
-        request_callback_resource_.get()));
+    // TODO: Support GPU memory.
+    *memory_type = TRITONSERVER_MEMORY_CPU;
+    *memory_type_id = 0;
+    return nullptr;
   }
-
-  static void PyTritonRequestReleaseCallback(
-      struct TRITONSERVER_InferenceRequest* request, const uint32_t flags,
-      void* userp)
+  static TRITONSERVER_Error* ResponseAllocatorBufferAttributesFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      struct TRITONSERVER_BufferAttributes* buffer_attributes, void* userp,
+      void* buffer_userp)
   {
-    py::gil_scoped_acquire gil;
-    auto cr = reinterpret_cast<CallbackResource*>(userp);
-    cr->release_fn(cr->request, flags, cr->user_object);
-    delete cr;
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryType(
+        buffer_attributes, TRITONSERVER_MEMORY_CPU));
+    ThrowIfError(TRITONSERVER_BufferAttributesSetMemoryTypeId(
+        buffer_attributes, 0 /* memory_type_id */));
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorAllocFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, const char* tensor_name,
+      size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id, void* userp, void** buffer, void** buffer_userp,
+      TRITONSERVER_MemoryType* actual_memory_type,
+      int64_t* actual_memory_type_id)
+  {
+    *buffer = malloc(byte_size * sizeof(uint8_t));
+    *actual_memory_type = TRITONSERVER_MEMORY_CPU;
+    *actual_memory_type_id = 0;
+    return nullptr;
+  }
+  static TRITONSERVER_Error* ResponseAllocatorReleaseFn(
+      struct TRITONSERVER_ResponseAllocator* allocator, void* buffer,
+      void* buffer_userp, size_t byte_size, TRITONSERVER_MemoryType memory_type,
+      int64_t memory_type_id)
+  {
+    free(buffer);
```

Comment:
can we add assert here to ensure memory type and memory type id are cpu and 0?

Reply:
yes, we can assert the memory type and id. [Assert memory type and id on release function](https://github.com/triton-inference-server/core/pull/426/commits/a8f7bda312521a7d3da57bf1d0e3483f87ea5d87)